part/set.go

```diff
@@ -141,6 +166,57 @@ func (s Set[T]) ToBytesFunc() func(T) []byte {
 	return s.toBytes
 }
 
+func (s Set[T]) MarshalJSON() ([]byte, error) {
+	if s.tree == nil {
+		return []byte("[]"), nil
+	}
+	var b bytes.Buffer
+	b.WriteRune('[')
+	iter := s.tree.Iterator()
+	_, v, ok := iter.Next()
+	for ok {
+		bs, err := json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+		b.Write(bs)
+		_, v, ok = iter.Next()
+		if ok {
+			b.WriteRune(',')
+		}
+	}
+	b.WriteRune(']')
+	return b.Bytes(), nil
+}
+
+func (s *Set[T]) UnmarshalJSON(data []byte) error {
+	dec := json.NewDecoder(bytes.NewReader(data))
+	t, err := dec.Token()
+	if err != nil {
+		return err
+	}
+	if d, ok := t.(json.Delim); !ok || d != '[' {
+		return fmt.Errorf("%T.UnmarshalJSON: expected '[' got %v", s, t)
+	}
+	for dec.More() {
+		var x T
+		err := dec.Decode(&x)
+		if err != nil {
+			return err
+		}
+		*s = s.Set(x)
```

Comment:
could be more efficient to gather the values and only then put them in a set; but not sure how much of a difference it makes

Reply:
It would allocate twice as much memory as needed. However we can optimize this by using a transaction so it won't do unnecessary copies of the inner nodes.